core/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -0,0 +1,177 @@
+        }
+        if (id != -1 && id >= max) {
+            throw new IllegalArgumentException("max must be greater than id");
+        }
+        this.max = max;
+        return this;
+    }
+
+    public SliceQuery toFilter(QueryShardContext context) {
+        final MappedFieldType type = context.fieldMapper(field);
+        if (type == null) {
+            throw new IllegalArgumentException("Field " + field + " not found");
+        }
+        if (UidFieldMapper.NAME.equals(field)) {
+            return new TermsSliceQuery(context.getIndexReader().getCoreCacheKey(), new Slice(field, id, max));
+        } else if (type.hasDocValues()) {
+            IndexFieldData<?> indexFieldData = context.getForField(type);
+            return new DocValuesSliceQuery(new Slice(field, id, max), indexFieldData);
+        }
+        throw new IllegalArgumentException("Can't slice on [" + field + "] because doc_values are not enabled");
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        innerToXContent(builder);
+        builder.endObject();
+        return builder;
+    }
+
+    void innerToXContent(XContentBuilder builder) throws IOException {
+        builder.field("field", field);
+        builder.field("id", id);
+        builder.field("max", max);
+    }
+
+    public static SliceBuilder fromXContent(QueryParseContext context) throws IOException {
+        XContentParser parser = context.parser();
+        XContentParser.Token token;
+        String currentFieldName = null;
+        String field = null;
+        int id = -1;
+        int max = -1;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token.isValue()) {
+                if (context.getParseFieldMatcher().match(currentFieldName, FIELD)) {
+                    field = parser.text();
+                } else if (context.getParseFieldMatcher().match(currentFieldName, ID)) {
+                    id = parser.intValue();
+                } else if (context.getParseFieldMatcher().match(currentFieldName, MAX)) {
+                    max = parser.intValue();
+                }
+            }
+        }
+        return new SliceBuilder().setField(field)
+            .setId(id).setMax(max);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (!(other instanceof SliceBuilder)) {
+            return false;
+        }
+
+        SliceBuilder o = (SliceBuilder) other;
+        return ((field == null && o.field == null) || field.equals(o.field))
+            && id == o.id && o.max == max;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.field, this.id, this.max);
+    }
+
+    @Override
+    public String toString() {
```

Comment:
if you do `extends ToXContentToBytes` instead of `implements ToXContent` you get that for free